### PR TITLE
feat(dashboard): Improve custom check editor

### PR DESCRIPTION
**Pull request by robertkasza**
## 🎯 Overview

Major refactor of the custom check editor in the database permissions system. This introduces a completely redesigned rule group architecture with better UX, enhanced functionality, and improved maintainability.

### New Rule Group Architecture (V2)
- **Tree-based Structure**: Replaced flat arrays with hierarchical tree nodes
- **Type Safety**: Discriminated union types for , , and
- **Enhanced Operators**: Added support for  logical operator
- **Native EXISTS Support**: First-class support for EXISTS subqueries (previously "unsupported")

### Improved User Experience
- **Visual Hierarchy**: Depth-based background colors for nested rule groups
- **Better Controls**: Intuitive add/remove buttons for rules, groups, and EXISTS clauses
- **Responsive Design**: Optimized layout for both desktop and mobile
- **Smart Validation**: Enhanced form validation with better error messaging

### Technical Improvements
- **New Parsing System**:  converts Hasura permissions to tree structure
- **New Serialization**:  converts back to Hasura format
- **Component Refactor**: Modular component architecture with , ,
- **Better Form Integration**: Deep React Hook Form integration with

## 📊 Code Changes

- **32 files changed**: +2,564 additions, -1,176 deletions
- **New Files**: 11 new components and utilities
- **Removed Files**: 4 legacy components and entire  directory
- **Tests**: Comprehensive test suite with 1,478+ lines of test coverage

## 🧪 Testing

- ✅ **Unit Tests**: Complete coverage for parsing/serialization logic
- ✅ **Component Tests**: React Testing Library tests for UI components
- ✅ **Integration Tests**: Form validation and data flow testing
- ✅ **Edge Cases**: Nested structures, EXISTS queries, operator combinations

## ⚠️ Breaking Changes

**Data Structure Migration Required**
- Old format:
- New format:

**API Changes**
- Removed:  utility
- Added:  and  functions

## 🛡️ Security & Performance

**Security Enhancements**
- UUID-based node identification
- Controlled component rendering
- Input sanitization through React forms

**Performance Considerations**
- Recursive operations optimized for reasonable nesting depths
- Memory-efficient UUID generation
- Form re-rendering optimizations

---

### Checklist

- [x] No breaking changes (data migration required)
- [x] Tests pass (comprehensive test suite included)
- [x] New features have new tests (full test coverage)
- [x] Documentation is updated (inline code documentation)
- [x] Title of the PR is in the correct format

### **PR Type**
Enhancement

___

### **Description**
- Replace the old `RuleGroupEditor` component with a new tree-based `CustomCheckEditor` that supports `_exists`, `_not`, `_implicit`, and relationship nodes natively (previously treated as "unsupported")
- Introduce a new `permissionUtils` module (`parsePermissionRule` and `ruleNodesToPermission`) for lossless round-trip conversion between Hasura permission JSON and an editable tree structure
- Add serialization collision validation (`findSerializationCollision`) to prevent users from creating permission structures that cannot be represented as valid Hasura JSON (e.g., duplicate conditions in implicit groups)
- Extend the `getAvailableOperators` utility to support `varchar` and `bpchar` column types alongside `text`
- Add depth-based visual nesting, disabled state propagation, and dark-mode support via new `secondary-{100..800}` Tailwind color tokens

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Hasura Permission JSON"] -->|parsePermissionRule| B["RuleNode[] tree"]
  B -->|wrapPermissionsInAGroup| C["GroupNode root"]
  C -->|CustomCheckEditor| D["React Hook Form state"]
  D -->|unWrapRuleNodes| E["Hasura Permission JSON"]
  subgraph UI Components
    D --> F["GroupNodeRenderer"]
    F --> G["ConditionRow"]
    F --> H["ExistsNodeRenderer"]
    F --> I["RelationshipNodeRenderer"]
    F -->|recursive| F
  end
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody>
<tr><td><strong>Core data model & serialization</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>types.ts</strong><dd><code>New discriminated union types for ConditionNode, GroupNode, ExistsNode, RelationshipNode</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-108bd951677d76732dc212eef899e4b326636982ebbe3696c97c1dba23ea30aa">+54/-0</a></td>
</tr>
<tr>
  <td><strong>parsePermissionRule.ts</strong><dd><code>New parser that converts Hasura permission JSON into a RuleNode tree</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-0fbd27547fa0634db5d5ec981512ac13016a182b5c84eff28865f886d93449d8">+197/-0</a></td>
</tr>
<tr>
  <td><strong>ruleNodesToPermission.ts</strong><dd><code>New serializer that converts RuleNode tree back to Hasura permission JSON</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-59cf5e8ac239eb4d95270aecd3e2d16e81c9e28ce9915c2647939228f9ff8d95">+129/-0</a></td>
</tr>
<tr>
  <td><strong>index.ts</strong><dd><code>Barrel exports for new permissionUtils module</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-9b527bf0a5ef615497113be1273685ae73bebf7f82983a905aa92e9e7fcf46c9">+19/-0</a></td>
</tr>
<tr>
  <td><strong>validationSchemas.ts</strong><dd><code>Rewrite Yup schemas for tree-based nodes; add findSerializationCollision checks</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-98f990165c3aca93bc01808ac0dcbde7b347ad2fd86fe52311d306a2fb3aaf0f">+141/-24</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>UI components (new)</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td><strong>CustomCheckEditor.tsx</strong><dd><code>Root editor component with empty-state handling and context provider</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-142cbdc65f09465703b0f9577ab113513e8ef2fba9a7ced7ba7ee47b8e7a48af">+97/-0</a></td>
</tr>
<tr>
  <td><strong>GroupNodeRenderer.tsx</strong><dd><code>Recursive renderer for group nodes with depth-based backgrounds</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-7ebb4460f85bd8c837eda10b2a6a74c42350d7e74f83b932c73cd2f0bc7cd881">+139/-0</a></td>
</tr>
<tr>
  <td><strong>ConditionRow.tsx</strong><dd><code>Row component for column/operator/value condition editing</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-f2c4497812d350a232054faa771ab084737a4df41bc0b80f86eb80838d9f5aec">+111/-0</a></td>
</tr>
<tr>
  <td><strong>ExistsNodeRenderer.tsx</strong><dd><code>Renderer for _exists nodes with table selector and nested where clause</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-7e61c7e4a4716fefb09f657817e481259bda4f38d7c9c0e3372023dad8edb0db">+83/-0</a></td>
</tr>
<tr>
  <td><strong>RelationshipNodeRenderer.tsx</strong><dd><code>Renderer for relationship nodes with dynamic context propagation</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-561252e77b7ded926114e5c05d668eca636986bf67fd58a7e5a769cb208283e0">+154/-0</a></td>
</tr>
<tr>
  <td><strong>LogicalOperatorBadge.tsx</strong><dd><code>Dropdown badge for switching between AND/OR/NOT/Implicit operators</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-85e63b6485b8adf4a1efaa5929e9a25ca758cf5120d2f23ae231ff12ba71e576">+95/-0</a></td>
</tr>
<tr>
  <td><strong>AddNodeButton.tsx</strong><dd><code>Popover button for adding conditions, groups, exists, and relationships</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-ffd4ef60cd6e57fa2ae50d9ac43bd8870ab6d28e1241cb0b14c9fae755e61e5d">+215/-0</a></td>
</tr>
<tr>
  <td><strong>TableComboBox.tsx</strong><dd><code>Searchable combobox for selecting schema.table in exists nodes</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-37badf2d5b6551d8803446122871efa030b23798b1ae05d241e3719e5eeb024c">+95/-0</a></td>
</tr>
<tr>
  <td><strong>RelationshipComboBox.tsx</strong><dd><code>Searchable combobox for selecting relationships</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-c3f86a2cdae37406aff54f4c0c0ae795cdb49287e1c8c3bd36d5dcc6384d60ea">+140/-0</a></td>
</tr>
<tr>
  <td><strong>useCustomCheckEditor.ts</strong><dd><code>Context hook replacing useRuleGroupEditor</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-a555da8d793d9308562a28825eb8e2dfdff22cf6abaa0611439d8b090540cade">+17/-0</a></td>
</tr>
<tr>
  <td><strong>index.ts</strong><dd><code>Barrel exports for CustomCheckEditor</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-47ddca44c2983e351d2734fcb9cd30b920eddce4eb5b8944c6a789282304c14e">+2/-0</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>UI components (renamed/updated)</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>ConditionValue.tsx</strong><dd><code>Renamed from RuleValueInput; switched to useCustomCheckEditor; added disabled prop passthrough</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-16890fffa8f090169d008c5b336bacfe0dc975dc98deb1247230d1181d7eb4bb">+11/-10</a></td>
</tr>
<tr>
  <td><strong>OperatorComboBox.tsx</strong><dd><code>Moved to CustomCheckEditor dir; switched from watch to useWatch; memoized maxOperatorLength</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-d8dba18cee6b0ebab115be994bff733e2adde01d56ae465bdde8b2ccaf825265">+8/-7</a></td>
</tr>
<tr>
  <td><strong>getAvailableOperators.ts</strong><dd><code>Extended text operators to varchar/bpchar; pre-computed operator arrays</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-2e87e9d9941dd38c7cdd6ca0a11eae17eb11cff6d659611522448b02c6bbd04b">+18/-8</a></td>
</tr>
<tr>
  <td><strong>ColumnAutocomplete.tsx</strong><dd><code>Removed useRuleGroupEditor dependency; added disabled prop; fixed any type</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-c89efa530042890e7d6277c2e3c763cb7c9b9fc1d7c14c62839f4cf7c42528f7">+6/-6</a></td>
</tr>
<tr>
  <td><strong>useAsyncValue.ts</strong><dd><code>Fixed optional chaining; added non-null assertion for column lookup</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-154398dc1685c03bb632fa8ad690d8d1a83a4376e9b5128cb1d6566b7ed51ec4">+3/-4</a></td>
</tr>
<tr>
  <td><strong>fancy-multi-select.tsx</strong><dd><code>Added disabled prop support with proper styling</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-6e6a7965a2c8e30e9a2c021a5009ff79e71de73892b86a3468d3474163dfeb03">+7/-2</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>Integration (form & permissions)</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>RolePermissionEditorForm.tsx</strong><dd><code>Switch from RuleGroup to GroupNode; use wrapPermissionsInAGroup/unWrapRuleNodes</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-c7cc670c499aaa76537a1ac3848721988fa0196d4cca8f6b5376b4a14f01341d">+16/-10</a></td>
</tr>
<tr>
  <td><strong>RowPermissionsSection.tsx</strong><dd><code>Replace RuleGroupEditor with CustomCheckEditor; update filter shape checks</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-663956d9adae1f6255151599b1cbd6ad03fea1246e87ab89329fcddcdbec2b20">+14/-13</a></td>
</tr>
<tr>
  <td><strong>dataBrowser.ts</strong><dd><code>Remove old RuleGroup interface (replaced by GroupNode type)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-33c6810dbd7e2910c86a15009467a348f064380b0e1dd787ef320b4e7543403b">+0/-12</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>Styling</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>globals.css</strong><dd><code>Add secondary-100 through secondary-800 CSS custom properties for light/dark</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-fe5e1172304066de64ab9008cce8cb7ac09661fbc93feb37d0ba69091f7f4311">+16/-0</a></td>
</tr>
<tr>
  <td><strong>tailwind.config.js</strong><dd><code>Register secondary-{100..800} Tailwind color tokens</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-0421515d64f36bf18988a5e62f6b406277d9a63b6991a8b3f4c9e976836449c8">+8/-0</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>Tests</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td><strong>CustomCheckEditor.test.tsx</strong><dd><code>Comprehensive tests for rendering, adding/removing nodes, operators, disabled state, edge cases</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-2b4e72005dc7f0c2b5a33cc54aa127adfbba70f54903aa28d9eab45aedcc32a3">+2233/-0</a></td>
</tr>
<tr>
  <td><strong>CustomCheckEditor.validation.test.tsx</strong><dd><code>Validation integration tests for empty columns, relationships, duplicates</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-d110c9fc4cbeccfa6ffdefe4ebfda9b3384d7e89d8bce800a55fe8f2045cf63c">+287/-0</a></td>
</tr>
<tr>
  <td><strong>parsePermissionRule.test.ts</strong><dd><code>Unit tests for parsing all Hasura permission patterns to RuleNode trees</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-9886d3291ecc2828974aa010f099da257f7e346c6b74fe9a635df3ee51017a9e">+697/-0</a></td>
</tr>
<tr>
  <td><strong>ruleNodesToPermission.test.ts</strong><dd><code>Round-trip tests ensuring lossless parse→serialize conversion</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-6e48d4808178f63dc42d233e9888c78d55f2fd18b98fe2709ce0b3e97272baa1">+403/-0</a></td>
</tr>
<tr>
  <td><strong>validationSchemas.test.ts</strong><dd><code>Tests for duplicate detection across conditions, operators, exists, relationships</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-7166f2b4e81b6435d69e4ae6a07c631ce52847567d958df6a5a7c4ae40a01fd7">+312/-0</a></td>
</tr>
<tr>
  <td><strong>RelationshipNodeRenderer.test.tsx</strong><dd><code>Tests for nested context propagation through relationship nodes</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-9c6b88c31b88160ca12bff71438f6167b5adeea8c5b14ddf5e3be1e6abaf4c1f">+226/-0</a></td>
</tr>
<tr>
  <td><strong>useAsyncValue.test.ts</strong><dd><code>Unit tests for async column value resolution hook</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-5180aad9d8afacb1b8494774e6ce6d8ce8525925ebbc255bead6a9b8c7be7af1">+177/-0</a></td>
</tr>
<tr>
  <td><strong>ConditionValue.test.tsx</strong><dd><code>Renamed from RuleValueInput.test; updated mocks</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-57be6a2c01c2e5d0a542df786960b526f0bbd77d30442377a05d8fb94289e5ee">+8/-8</a></td>
</tr>
<tr>
  <td><strong>getAvailableOperators.test.ts</strong><dd><code>Extended tests for varchar/bpchar column types</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-2e9c5372798b259b3cdec1ce477319111cd2fb3ab4d90d23d9da76a32fe0bace">+10/-5</a></td>
</tr>
<tr>
  <td><strong>RowPermissionSection.test.tsx</strong><dd><code>Updated tests for new CustomCheckEditor UI and GroupNode data shape</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-2a32fbb9eda12ec8eb93746c5c8b171e8ae20d18e661a5e2eb0c4996fee8376b">+34/-31</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>E2E tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>permissions-graphql-verification.test.ts</strong><dd><code>New e2e test verifying custom check permissions work end-to-end with GraphQL</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-03238bf6d86a7e8899c0902625adc616094300fac734fb6141403fd47474d179">+149/-0</a></td>
</tr>
<tr>
  <td><strong>permissions-table.test.ts</strong><dd><code>Update e2e selectors for renamed "Add check" button</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-fd25341f9701802f0da5a821f0156661c2bc98bd49ed2c49bfd8ba7b278bbd0c">+2/-3</a></td>
</tr>
<tr>
  <td><strong>utils.ts</strong><dd><code>Add e2e helpers for SQL editor, GraphQL playground, headers, query execution</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-490448aa83585151d8c61d698273c43486fdcac6a5d28a9b7e5be2729bbffd12">+77/-0</a></td>
</tr>
</table></details></td></tr>
<tr><td><strong>Deleted (old implementation)</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td><strong>RuleGroupEditor.tsx</strong><dd><code>Removed old permission rule editor component</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-967b5885eecb27e94308f351d7e1e9f563b8e1f593ac9cfec9efc7d155604ef4">+0/-271</a></td>
</tr>
<tr>
  <td><strong>RuleEditorRow.tsx</strong><dd><code>Removed old rule editor row (replaced by ConditionRow)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-a7a1d2aa882735a2b9cfb41e95b05c6777d706570eec5deec6bf5d2381a51252">+0/-121</a></td>
</tr>
<tr>
  <td><strong>RuleGroupControls.tsx</strong><dd><code>Removed old group controls (replaced by LogicalOperatorBadge)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-dc0bc7370cec4b048fb5d43bae46fb94a7490fa8bf2258fca77598169769c8b9">+0/-78</a></td>
</tr>
<tr>
  <td><strong>RuleRemoveButton.tsx</strong><dd><code>Removed old remove button (replaced by inline trash buttons)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-022b63a861fefcbbb471c62bfe44a020f11284487d2a6e551815b1ae69d47222">+0/-53</a></td>
</tr>
<tr>
  <td><strong>useRuleGroupEditor.ts</strong><dd><code>Removed old context hook (replaced by useCustomCheckEditor)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-ad47089637abb03e832dd7f40642cd9ecd550c3814cbefa6558aac8cf55c7db6">+0/-21</a></td>
</tr>
<tr>
  <td><strong>RuleGroupEditor/index.ts</strong><dd><code>Removed old barrel export</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-76adc6b3bcf3f11960a496849ff2f578fcf819d71357c6ad05966baf0d2f590f">+0/-2</a></td>
</tr>
<tr>
  <td><strong>convertToHasuraPermissions.ts</strong><dd><code>Removed old serializer (replaced by ruleNodesToPermission)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-046bb93fc9fd9abd712719cd01982ebe633596af1e3ca488403d22a32c2c067e">+0/-102</a></td>
</tr>
<tr>
  <td><strong>convertToHasuraPermissions.test.ts</strong><dd><code>Removed old serializer tests</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-8fe2760893d98461d9c3cfbabf6dc8b650c48bc1b8f5bb4fe93888a4b4f9da2b">+0/-297</a></td>
</tr>
<tr>
  <td><strong>convertToRuleGroup.ts</strong><dd><code>Removed old parser (replaced by parsePermissionRule)</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-299154e148e3401d15b604b2f52b67229b2a2696e02b9c89d4a0c035110fbb04">+0/-312</a></td>
</tr>
<tr>
  <td><strong>convertToRuleGroup.test.ts</strong><dd><code>Removed old parser tests</code></dd></td>
  <td><a href="https://github.com/nhost/nhost/pull/3931/files#diff-7e1b7b41b64c53e3b8c5dca48fbba781cfbe813936a45643a174af8dd0d6c6e0">+0/-470</a></td>
</tr>
</table></details></td></tr>
</tbody></table>

</details>

___

### CODE REVIEW

File: dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/ConditionRow.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+      />
+
+      <Button
+        variant="ghost"
+        size="icon"
+        className="w-full xl:w-auto"
+        onClick={onRemove}
+        disabled={disabled || disableRemove}
+      >
+        <Trash2 className="h-4 w-4" />
```
Comment (dbm03) [inline dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/ConditionRow.tsx:108]: nit: add `text-destructive` when it's not disabled. That way it's consistent with other trash icon buttons we have in the dashboard

Image

File: dashboard/tailwind.config.js [hunk 2]
```diff
@@ -25,6 +25,7 @@ module.exports = {
         'data-cell-bg-hover': 'hsl(var(--data-cell-bg-hover))',
         'data-cell-bg-disabled': 'hsl(var(--data-cell-bg-disabled))',
         'data-table-border-color': 'hsl(var(--data-table-border-color))',
+        'check-permissions-bg': 'hsl(var(--check-permissions-bg))',
```
Comment (dbm03) [inline dashboard/tailwind.config.js]: we're not using this variable, maybe it's a leftover (unless you planned to use it in the future)

File: dashboard/src/styles/globals.css [hunk 3]
```diff
@@ -315,6 +324,15 @@ body {
     --default: 216.92 23.64% 10.78%;
     --primary-highlight: 215.88 100% 60.98% / 16%;
     --error-toast-bg: 221.5, 10.7%, 23.7%;
+    --check-permissions-bg: 215 15.4% 15.3%;
```
Comment (dbm03) [inline dashboard/src/styles/globals.css]: maybe leftover variable

File: dashboard/src/styles/globals.css [hunk 4]
```diff
@@ -272,6 +272,15 @@ body {
     --default: 0 0% 98.04%;
     --primary-highlight: 216 100% 40.2% / 8%;
     --error-toast-bg: 215.7 12.9% 70.6%;
+    --check-permissions-bg: 204, 29.4%, 96.7%;
```
Comment (dbm03) [inline dashboard/src/styles/globals.css]: maybe leftover variable

File: dashboard/src/features/orgs/projects/database/dataGrid/utils/permissionUtils/ruleNodesToPermission.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      sourceVal !== null &&
+      !Array.isArray(sourceVal)
+    ) {
+      result[key] = deepMerge(
+        targetVal as Record<string, unknown>,
+        sourceVal as Record<string, unknown>,
+      );
+    } else if (Array.isArray(targetVal) && Array.isArray(sourceVal)) {
+      console.error(
+        `deepMerge: array collision on key "${key}", keeping original value`,
```
Comment (dbm03) [inline dashboard/src/features/orgs/projects/database/dataGrid/utils/permissionUtils/ruleNodesToPermission.ts]: I'm getting this silent error in console when I click on Save with the following form values:

Image

I think it has to do with the fact that we have two different values with the same key, such as `"_and": {}, "_and": {}`.

The problem is that the UI makes me think that everything's fine. The error message only shows in the console. 

Image

But actually, I see that I've lost some of the data the next time I open the same section:

Image


P.S. To be honest, I don't know if this exact permission settings would make any sense in a real project use-case. But better error handling could be nice I guess

File: dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/CustomCheckEditor.validation.test.tsx
Comment (dbm03) [inline dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/CustomCheckEditor.validation.test.tsx:1]: Why the `.validation` in the filename? Just curious, since I haven't seen it in other place

File: dashboard/e2e/utils.ts [hunk 6]
```diff
@@ -536,6 +536,83 @@ export async function createRelationship({
...
+  await page.getByRole('button', { name: 'Execute GraphQL query' }).click();
+}
+
+export async function getGraphQLResult({
+  page,
+}: {
+  page: Page;
+}): Promise<string> {
+  const resultWindow = page.getByLabel('Result Window');
+  await page.waitForTimeout(2000);
```
Comment (dbm03) [inline dashboard/e2e/utils.ts]: Instead of a hardcoded timeout, why can't we use a condition-based wait?

Maybe, something like:

```ts
await expect(resultWindow).not.toHaveText('');
```

I think we're just waiting for the result window to not be empty, aren't we?

File: dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/CustomCheckEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        type: 'group',
+        id: uuidv4(),
+        operator: '_implicit',
+        children: [node],
+      });
+    }
+  }
+
+  function handleRootRemove() {
+    setValue(name, { type: 'group', operator: '_implicit', children: [] });
```
Comment (dbm03) [inline dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/CustomCheckEditor.tsx]: nit: missing `id: uuidv4()`

File: dashboard/src/features/orgs/projects/database/dataGrid/components/EditPermissionsForm/sections/RowPermissionsSection.tsx [hunk 8]
```diff
@@ -51,9 +53,7 @@ export default function RowPermissionsSection({
   const { filter } = getValues();
 
   const defaultRowCheckType =
-    isNotEmptyValue(filter?.rules) ||
-    isNotEmptyValue(filter?.groups) ||
-    isNotEmptyValue(filter?.unsupported)
+    isNotEmptyValue(filter?.children) || isNotEmptyValue(filter?.unsupported)
```
Comment (dbm03) [inline dashboard/src/features/orgs/projects/database/dataGrid/components/EditPermissionsForm/sections/RowPermissionsSection.tsx]: I believe the second check is dead code: `filter?.unsupported`

`unsupported` was only valid with the old interface RuleGroup

```ts
export interface RuleGroup {
  operator: '_and' | '_or';
  rules: Rule[];
  groups: RuleGroup[];
  // biome-ignore lint/suspicious/noExplicitAny: TODO
  unsupported?: Record<string, any>[];
}
```

File: dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/CustomCheckEditor.validation.test.tsx
Comment (robertkasza) [reply dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/CustomCheckEditor.validation.test.tsx:1]: So, the CustomCheckEditor.test.tsx was getting to big and just moved in to a different file. But I've deleted the file because realized that it will not catch if we break validation somehow.

And also refactored  the customCheckEditor test because mocking the component ended up misleading and test were passing because we were mocking components.

File: dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/test-cases.md
Comment (dbm03) [inline dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/test-cases.md:1]: This is perhaps a leftover, unless you want to keep it as documentation?

Comment (robertkasza) [reply dashboard/src/features/orgs/projects/database/dataGrid/components/CustomCheckEditor/test-cases.md:1]: This is just a leftover, I've removed it. Thanks
